### PR TITLE
Fix zero initialization of workgroup memory

### PR DESCRIPTION
**Pull request by teoxoy**
Use local (not global) invocation id - duh 🙄.
Credits go to @nical who found out the issue!

### CODE REVIEW

File: src/back/hlsl/writer.rs [hunk 1]
```diff
@@ -1176,10 +1176,7 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
                         if !func.arguments.is_empty() {
                             write!(self.out, ", ")?;
                         }
-                        write!(
-                            self.out,
-                            "uint3 __global_invocation_id : SV_DispatchThreadID"
-                        )?;
+                        write!(self.out, "uint3 __global_invocation_id : SV_GroupThreadID")?;
```
Comment (jimblandy) [inline src/back/hlsl/writer.rs:1179]: We probably want to change the HLSL variable name as well. It is no longer a "global invocation id".

Pre-existing, but: What happens if the entry point already has an argument that is the `BuiltIn::LocalInvocationId`? Can HLSL handle having two `SV_GroupThreadId` arguments on the same function?

File: src/back/msl/writer.rs [hunk 2]
```diff
@@ -3695,7 +3695,7 @@ impl<W: Write> Writer<W> {
                 };
                 writeln!(
                     self.out,
-                    "{separator} {NAMESPACE}::uint3 __global_invocation_id [[thread_position_in_grid]]"
+                    "{separator} {NAMESPACE}::uint3 __global_invocation_id [[thread_position_in_threadgroup]]"
```
Comment (jimblandy) [inline src/back/msl/writer.rs:3698]: Similarly, let's rename this `__local_invocation_id`.

File: src/back/msl/writer.rs [hunk 3]
```diff
@@ -3659,7 +3659,7 @@ impl<W: Write> Writer<W> {
                     &self.names[name_key]
                 };
 
-                if binding == &crate::Binding::BuiltIn(crate::BuiltIn::GlobalInvocationId) {
+                if binding == &crate::Binding::BuiltIn(crate::BuiltIn::LocalInvocationId) {
```
Comment (jimblandy) [inline src/back/msl/writer.rs:3662]: Let's rename the Rust variable to `local_invocation_id`.

**Review by jimblandy — CHANGES_REQUESTED** [review]
Sorry I missed this in the first review.

File: src/back/hlsl/writer.rs (hunk 1 above)
Comment (teoxoy) [reply src/back/hlsl/writer.rs:1179]: > Pre-existing, but: What happens if the entry point already has an argument that is the `BuiltIn::LocalInvocationId`? Can HLSL handle having two `SV_GroupThreadId` arguments on the same function?

Yeah, only msl and spv seem to complain.

**Review by jimblandy — APPROVED** [review]

File: src/back/spv/writer.rs [hunk 4]
```diff
@@ -430,7 +430,7 @@ impl Writer {
                         constituent_ids.push(id);
 
                         if binding == &crate::Binding::BuiltIn(crate::BuiltIn::GlobalInvocationId) {
```
Comment (DJMcNab) [inline src/back/spv/writer.rs:432]: I'm digging into this again for gfx-rs/wgpu#4592. Why was this one left as Global?

Comment (teoxoy) [reply src/back/spv/writer.rs:432]: It seems I missed this one 😅

**Pull request merged**